src/testSession.ts

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+import * as path from 'path';
+import * as os from 'os';
+import { debug, Debugger } from 'debug';
+import { fs as fsCore } from '@salesforce/core';
+import { env, parseJson } from '@salesforce/kit';
+import { getString, Optional } from '@salesforce/ts-types';
+import { createSandbox, SinonStub } from 'sinon';
+import * as shell from 'shelljs';
+import { genUniqueString } from './genUniqueString';
+import { zipDir } from './zip';
+import { TestProject, TestProjectConfig } from './testProject';
+
+export interface TestSessionOptions {
+  /**
+   * Specify a different location for the test session.
+   */
+  sessionDir?: string;
+
+  /**
+   * Define a test project to use for tests.
+   */
+  project?: TestProjectConfig;
+
+  /**
+   * Commands to run as setup for tests.  All must have exitCode == 0 or session
+   * creation will throw.  If an org create command is specified the test session
+   * will have a `orgUsername` property for use in tests.
+   */
+  setupCommands?: string[];
+}
+
+// Executes commands and sets the org username if an org was created.
+// Throws if any commands return a non-zero exitCode.
+const setupCommands = (testSession: TestSession, cmds?: string[]): void => {
+  if (cmds) {
+    const dbug = debug('testkit:setupCommands');
+    for (let cmd of cmds) {
+      // if it's an org:create without --json, append it anyway so we
+      // can get the org username and keep track of it.
+      if (cmd.includes('org:create')) {
+        if (testSession.orgUsername) {
+          return;
```

Comment:
Setup commands allows for an arbitrary set of sfdx commands to be executed, yet does not provide any means of capturing data that the test writer might need later. 

Could we consider providing a way for the test writer to "hook" into the setup code with a callback that would allow the test writer to examine the command and its results and give them the ability to store data into the test session. These data can then be used by test writer later, when needed.

Reply:
Both excellent points Pete - thanks!  I'll fix the first issue now.  The 2nd I'll think of something.  Probably just add it to the test session as a `Map<string, ShellString>` where key=cmd and value=cmd_results.